Search: Add MockUsageMeter along with DonutMeterContainer component

a8ck3n opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #26052

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Introduce `DonutMeterContainer` component.
* Introduce `MockUsageMeter` to Jetpack Search Dashboard.
* Use pricing gate `isNewPricing202208` to display the `MockUsageMeter` component.

> Note by @dognose24: Since this DonutMeterContainer component may stay inside the Search package, I would prefer to complete the whole MockUsageMeter along with the DonutMeterContainer component to avoid redundant PR processes.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Per discussion in Slack, we're going to build this directly in the Search package for now and look at Storybook integration as a follow up. It looks like we can build components as part of our package and still expose them to Storybook.

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

N/A

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Navigate the site URL with `/wp-admin/admin.php?page=jetpack-search&new_pricing_202208=1`.
* Ensure there will be a `UsageMeter` section displayed.
* Ensure the `UsageMeter` will be in line with the design.

| Before | After |
| --- | ----------- |
| <img width="1228" alt="截圖 2022-10-03 上午9 36 08" src="https://user-images.githubusercontent.com/6869813/193487221-7f84818f-9b44-46b0-9b11-5cb590b2211c.png"> | <img width="1280" alt="截圖 2022-10-03 上午9 35 39" src="https://user-images.githubusercontent.com/6869813/193487234-466edf71-6807-4849-a51c-3d9b5a18d29d.png"> |

a8ck3n committed:
Export DonutMeter from component library.

a8ck3n committed:
First pass at unstyled output.

Generates the necessary HTML for the component.

a8ck3n committed:
First pass at styling the component.

a8ck3n committed:
Move component code into separate function.

Now it’s a component! We’re getting there…

a8ck3n committed:
Add changelog.

a8ck3n committed:
Added changelog for component library.

oskosk commented on projects/packages/search/src/dashboard/components/pages/dashboard-page.jsx:
```diff
@@ -147,13 +168,29 @@ export default function DashboardPage( { isLoading = false } ) {
 			{ isPageLoading && <Loading /> }
 			{ ! isPageLoading && (
 				<div className="jp-search-dashboard-page">
 					{ renderHeader() }
 					{ renderMockedSearchInterface() }
+					{ renderDonutMeterTestSection() }
```
Let's not do this. There's no point in using functions called `renderSomething` when components are actually functions that just return what's gonna be rendered. If you go with this approach this function is not visible for the React Dev Tools as a component. 

Go for a name like `const DonutMeterTestSection = ()...` and use it like this `<DonutMeterTestSection />`

In this particular example I understand there were `renderHeader` and `renderMockedSearchInterface` not introduced in the PR, but let's just do it properly with `DonutMeterTestSection`

a8ck3n replied:
Thanks for the feedback @oskosk! Much appreciated!

In this case, I'm building out `DonutMeterWithLabels` as the actual component but I understand what you're getting at here. Maybe moving this code block into a "layout" component is the best option here? I'll give that a try as I continue to build this out.

a8ck3n replied:
Thanks for the feedback @oskosk! Much appreciated!

In this case, I'm building out DonutMeterWithLabels as the actual component but I understand what you're getting at here. Maybe moving this code block into a "layout" component is the best option here? I'll give that a try as I continue to build this out.

a8ck3n committed:
Fixes for linting errors.

a8ck3n committed:
New files for DonutMeterContainer.

Now inside “components/donut-meter-container” where it belongs.

a8ck3n committed:
Introduce some sub-compenents.

a8ck3n committed:
Build out InfoSecondary component.

Added props for the message string and the callback function to be called when the user clicks the link.

a8ck3n committed:
Build out InfoPrimary component.

Added props for the message string and the callback function to be called when the user clicks the icon.

a8ck3n committed:
Refactor handling of click events.

oskosk committed:
Merge branch 'trunk' into add/donut-meter-container

oskosk committed:
.

oskosk commented:
Pushed a rebase in [171f874](https://github.com/Automattic/jetpack/pull/26344/commits/171f874d1684404d87117c3cf0ea9e5976581eff) and [7c04d95](https://github.com/Automattic/jetpack/pull/26344/commits/7c04d95c24cdf56d07a90eacffbfca89f224c2ed)

kangzj commented:
I'll shepherd this further after I mock the required APIs and if no one gets on it first. 🐑

dognose24 committed:
Add usage info to DonutMeterContainer

dognose24 committed:
Add i18n to DonutMeterContainer and title

dognose24 committed:
Introduce .upgrade-trigger section

dognose24 committed:
Add .record-meter-about section

dognose24 committed:
Refactor .jp-search-dashboard-meter-wrap styles

dognose24 committed:
Merge branch 'trunk' into add/donut-meter-container

dognose24 committed:
Add the pricing gate to UsageMeter component

dognose24 committed:
Update changelog

kangzj committed:
refactor numberWithCommas

kangzj reviewed (approved):
Looks good to me 👍 

I tweaked the number formatting a bit because some countries use dot as thousand separators rather than commas.

Another thing I noticed is that the donut meters don't change colors with the usage, like if the usage is above 100%, then the meter should be red.

Thanks for shepherding this @dognose24. Great job! 👍

dognose24 commented:
> I tweaked the number formatting a bit because some countries use dot as thousand separators rather than commas.

Thanks for the more general solution!

> Another thing I noticed is that the donut meters don't change colors with the usage, like if the usage is above 100%, then the meter should be red.

Sure thing! There should be another PR to address donut colors and event callbacks.
